### ISSUE TITLE
Feature request - area for custom content

### ISSUE THREAD

**Issue by weilbith** [akinsho/nvim-bufferline.lua#98]
Hey 👋🏾 

your plugin looks very appealing and has really nice features. 👍🏾 

I really like to have such a nice tab-styled bufferline. But I also want to have some additional global information shown there. For example I would like to try moving the git branch info from the statusline to the bufferline because it won't change per buffer/window but is a more global information (for my use-cases). Another example would be the status of the language server or number of global workspace diagnostics. I actually would love to have a unified or second statusline at the bottom of NeoVim, but that does not exist.
I don't want to ask you to support all these features. Rather I was thinking if it wouldn't be a nice generic feature to let the user have a reserved space on the right side of the bufferline for his own content. So there would be two options. One for the guaranteed space this segment gets and one for a function which returns the content of this segment. Interesting would be to have different autocommand events which would trigger to re-evaluate the function to update the segment, but that's maybe too much at once. The buffer "tabs" should then just truncate as now. Maybe the guaranteed size is not even necessary and could be more dynamic. But that sounds more complicated and would make a first version of this feature way harder (I assume).

What do you think about this idea in general? 🙃 
I think it is kinda cool, because it allows to use your fancy UI parts and combine if a quite generic interface to become extended with some additional useful information. Looking forward to hear what you have to say. 🤗

**Comment by akinsho** [comment]
@weilbith thanks for the kind words and checking the plugin out.

Tbh in general I think the scope of this plugin/complexity is increasing too much, I've added in *way* more than I planned so I'm a bit reluctant especially since as you pointed out this is a bit of a workaround for the fact that nvim doesn't support a global statusline. Tbh I'd love to see that investigated in nvim rather than working around that sort of thing here.

All that being said I've seen a few people already try and do this with this plugin using some workarounds like getting the string from the tabline function and appending to it. 

Tbh if someone else was willing to put in the work I wouldn't refuse a PR since people seem to want this. I think regarding implementation there are some tricky things to consider.

A static size will be a nightmare since I think this will quickly become an issue magnet if people put things with variable size in there like lsp status so will be hard for them to predict.

I think the best way to implement this will be similar to how the `offset` feature works, a user specifies something like
```lua
config = {
  custom_section = {
    left = {
      {text = "<icon>", highlight = {guifg = "<gui>", guibg = "<guibg>", gui = "<gui>"}
      {text = "<text>"}
    }
}

-- then in the plugin those things can be concatenated into a valid string e.g.
local result = ""
local total_size = 0
for _, item in (custom_section.left) do
  total_size = total_size + #item.text
  -- this will output something like %#HighlightName#text_from_user
  result = result .. require('highlights').hl(generate_name()) .. item.text
end
```

This can then be used a bit like
https://github.com/akinsho/nvim-bufferline.lua/blob/64ba179ea810b868eda1031b2c476596657e3a52/lua/bufferline.lua#L585

Regarding how to update the tabline, tbh it updates natively very frequently so I don't know how often that will be an issue, I'd rather see exactly how this problem comes up rather than start in with a bunch of refreshing autocommands or something especially since it will cause the whole statusline to need to be recalculated which could become excessive

Anyway in summary I'd accept a PR for this and ☝🏾 is a good way to go but I'm unlikely to implement this myself since I'd rather avoid the complexity/I'm unlikely to ever use it personally but since there's demand I'd be alright with a PR.

**Comment by weilbith** [comment]
Cool, thank you very much for your detailed answer. I highly appreciate it.
I can totally follow your arguments and agree with them. The related [NeoVim issue](https://github.com/neovim/neovim/issues/9342) hasn't seen an update since ages. Apparently there is not enough support for it. So I guess I have to decide what I want.
To be honest I probably won't open a PR myself. Therefore it's too much braided into a plugin I don't know and the usage gain is medium while the complexity of the plugin increases. But we will see. Maybe some day I wake up super motivated to implement it.

**Comment by weilbith** [comment]
Do you want to close the issue and re-open it when it becomes relevant again ?

**Comment by akinsho** [comment]
@weilbith you can leave it open, sometimes despite my initial protestation I decide to implement these features. It's not really very complicated to actually implement it's just more that the number of things this is now doing is growing in a way that I fear one day I will wake up and be too overwhelmed to keep maintaining it 😆 (we're not really close to that yet tbh though)

**Comment by weilbith** [comment]
Haha. I totally understand.
Actually thats why I wanted it to be so generic so you don't have to maintain tons of new stuff, but just this relatively easy feature. But sure, generic often means more responsibility and a wider field of use-cases which increases the possibility to break on edge-cases.
